### ISSUE TITLE
[1.0] Graph queries do not work from components outside /pages

### ISSUE THREAD

**Issue by iiroj** [gatsbyjs/gatsby#553]
It seems when I have a component inside `${__dirname}/components`, I cannot use graph queries in it. For example, cloning [your blog](https://github.com/KyleAMathews/blog) and abstracting the blog post list into its own component required me passing it `props.data.allMarkdown` from _index.js_, instead of being able to query inside it directly.

From taking a quick look at `query-runner.js` it seems you can only run queries in routes, not components. Maybe this should be taken into account?

**Comment by KyleAMathews** [comment]
Yes — this is a problem. Quite often a component would want to have its own query but as you've noticed, they can't.

I have two plans for improving this.

First is @andreypopp pointed out a sneaky way to use Webpack loaders w/ graphql queries. You can require an empty resource while passing in a dynamic query something like:

```javascript
require(`graphql?query=${encodeURIComponent(`
{
  allMarkdown {
    edges {
      node {
        path
      }
    }
  }
}
`)
}!`)
```

Going to prototype this soon. Probably will just create a wrapper function to hide the weird syntax. This will let components query for whatever data they want.

I would say let's just use this for everything but this doesn't work for template page components. E.g. for a blog you have a blog post component that's used for every blog post.

To make queries work in this scenario, there needs to be a way to pass in page-specific variables (e.g. the path). The solution I came up with is is that the graphql query is exported and then run multiple times, once for each page the template component is used for. E.g. this query for blog posts from my blog https://github.com/KyleAMathews/blog/blob/90e9c2979b54f08b76ddd7e9f7101b6ba93d6353/page-templates/blog-post.js#L108

This also complicates querying for data for any component used in template components as their data is also (sometimes) page specific. E.g. the data for my `ReadNext` component is dependent on the page it's on. The only solution I see to that is that components export their query and that gets composed within the parent component's query and then the parent passes down the necessary data into the child. This is what Relay works and while it's a bit ungainly, it works just fine. This btw, isn't supported quite yet. The query-runner.js doesn't know how to follow imports yet back to their source.

So tl;dr for graphql queries is always use the (coming) requireGraphQL method except for template pages where you export a query.

Thoughts?

**Comment by iiroj** [comment]
This seems logical. Having to declare that a component wants to make a query is at least verbose. Maybe it'll even improve performance.

With this in mind though, if  I have "have to" import some GraphQL method into my _BlogPostList_ component anyway, I wouldn't mind having to do the same for my _BlogPost_ layout component.

**Comment by MarcCoet** [comment]
Any update on this side

**Comment by fabien0102** [comment]
Otherwise you can export a query string in the component, and recompose queries into page

I think this solution can be clean ;)

**Comment by KyleAMathews** [comment]
@fabien0102 that actually doesn't work yet. We need to improve our Babel parser to support getting strings from other pages. Totally doable just hasn't been done yet.

I agree this is a very elegant pattern and what the various GraphQL clients seem to be settling on. Each component can export a query fragment and then those get composed into the page query. This blog posts talks about how Relay Modern does things https://dev-blog.apollodata.com/exploring-relay-modern-276f5965f827

The webpack loader idea is still intriguing to me though limited as it doesn't have access to per-page context variables (e.g. which blog post am I on) so can only be used for getting global data.

**Comment by fabien0102** [comment]
Very strange, strings import/exports works perfectly for value into templates but it's totally broken if you try to compose a query with import string… Now, I understand this issue, and it's looks a tricky bug!

**Comment by KyleAMathews** [comment]
We just need to add code to [this file](https://github.com/gatsbyjs/gatsby/blob/1.0/packages/gatsby/lib/query-runner/query-extractor.js) which detects if a variable came from another file and then to parse that file and get the exported data.

**Comment by fabien0102** [comment]
Other idea, add a graphql loader into webpack -> http://dev.apollodata.com/react/webpack.html (I never play with this, it's just an idea ;) )

**Comment by MarcCoet** [comment]
If I am not mistaken, we can't query from layouts either... Would it be easier to implement than from components?

**Comment by pinaudnico** [comment]
Yes I query from layout, but how can I have my result of query in a component ?

**Comment by DaleWebb** [comment]
> Yes I query from layout, but how can I have my result of query in a component ?

I normally pass the result of the query through the props of the component that needs it. Unless you want to be able to access the result in any component that is rendered?

**Comment by pinaudnico** [comment]
First thanks for the help. But I'm not sure to understand, in my layout, index.js I have something like : <div > {children()} </div>, and if I understand I have to pass the result of the query to this "children", how to write that ? If I would access from any of the component how to do ? Thanks of a real beginner.

**Comment by DaleWebb** [comment]
So `children()` is the React Element that will be rendered. As results from GraphQL queries are normally populated through the `props` attribute, you could pipe your results though there.

For example:

```javascript
export const pageQuery = graphql`
  query navigation {
    site {
      siteMetadata {
        url
      }
    }`

...

render() {
  let children = this.props.children();
  children.props.data.navigation = this.props.data.navigation;
  return (
    <div>
      {children}
    </div>
  );
}
```

Only the immediate child will have the results of the query in its `props` attribute, subsequent children won't.

This feels like a bit of a hack and I don't know if it's good practice, but see if it works for you.

**Comment by pinaudnico** [comment]
here is my code 


export default ({ data, children }) => {
  // la je peux mettre du js pur
  console.log (data);  
  
  
  let children = this.props.children();
  children.props.data.navigation = this.props.data.Query;             
          
  return (        
          <div>
               
                      <div>
                              <div>Coordonnées du point enregistré</div>
                              <div>Latitude : {data.allTableaudespointsCsv.edges[0].node.field2}</div>
                              <div>Longitude : {data.allTableaudespointsCsv.edges[0].node.field3}</div>
                              <br />
            
                              <div>Coordonnées actuelles</div>
                              <GPS />
            
            
                              <div>
                              {children()}
                              </div>
                      </div>
           </div>  
     )
}


export const query = graphql`
  query Query {
  allTableaudespointsCsv {
    edges {
      node {
        field1
        field2
        field3
        field4
        field5
        field6
      }
    }
  }
  
}
`

it says me Module build failed: Duplicate declaration "children" on the line with the let word.

**Comment by DaleWebb** [comment]
Change the name of the `children` variable to something else on this line: `let children = this.props.children();`

**Comment by pinaudnico** [comment]
I've changed in 

export default ({ data, boy }) => {
// la je peux mettre du js pur
console.log (data);

let boy = this.props.children();
boy.props.data.navigation = this.props.data.requete;

Duplicate declaration seems to come from my line export default ({ data, boy }) => {
but I need to export ???

**Comment by MarcCoet** [comment]
@pinaudnico you are destructuring the props. To fix the error, you need to do something like
```
export default ({ data, children }) => {
// la je peux mettre du js pur
console.log (data);

let boy = children();
boy.props.data = data
```
because in your code `props === { data, children, ...rest }`
If this does not ring a bell, you should read about react and about destructuring in JS

Personally I would rewrite your code like
```
export default (props) => {
// la je peux mettre du js pur
console.log (props.data)

return (
                  <div>
                          <div>Coordonnées du point enregistré</div>
                          <div>Latitude : {data.allTableaudespointsCsv.edges[0].node.field2}</div>
                          <div>Longitude : {data.allTableaudespointsCsv.edges[0].node.field3}</div>
                          <br />
        
                          <div>Coordonnées actuelles</div>
                          <GPS />
        
        
                          <div>
                          {children({
                              data: props.data
                          })}
                          </div>
                  </div>
       </div>  
   )
}
```
Then in the children aka the pages you will find `props.data`

BUT my guess is you could do things another way because this approach has its flaws.

**Comment by pinaudnico** [comment]
I've got an error : children is not defined

**Comment by fabien0102** [comment]
The doc to understand functionnal/classic react component -> https://reactjs.org/docs/components-and-props.html#functional-and-class-components

To resume:
```js
// Class component
class MyComp extends React.Component {
  render() {
    console.log(this.props); // {a: "toto", b: "plop"}
    return <div>{this.props.a}</div>;
  }
}

// Functionnal component (without decomposition)
const MyComp = props => {
  console.log(props); // {a: "toto", b: "plop"}
  return <div>{props.a}</div>;
};

// Functionnal component (with decomposition)
const MyComp = ({ a, b }) => {
  console.log(a, b); // "toto", "plop"
  return <div>{a}</div>;
};

// Usage
const ParentComp = () => <MyComp a="toto" b="plop" />
```

So with your example:

```js
export default ({ data, children }) => {
  console.log(data, children);

  return (
    <div>
      <div>Coordonnées du point enregistré</div>
      <div>Latitude : {data.allTableaudespointsCsv.edges[0].node.field2}</div>
      <div>Longitude : {data.allTableaudespointsCsv.edges[0].node.field3}</div>
      <br />

      <div>Coordonnées actuelles</div>
      <GPS />

      <div>
        {children({ data })}
      </div>
    </div>
  );
};
```

BUT you can't give `data` to your children like that ^^ (children is not a function), you need create new react component with props. This is a little codepen to illustrate the trick (lines 4-8): https://codepen.io/fabien0102/pen/QOaBLy?editors=0010

Good luck ;)

**Comment by MarcCoet** [comment]
@pinaudnico my bad. It is `props.children(...)` and `props.data. ...` in the render method. I was writing too fast.
`children` is indeed a function in the layout but there are some shortcomings. For example, last time I checked, if you changed `data`, the children would not rerender (like normal components receiving props would do). So it is only useful for passing data on mount.

**Comment by pinaudnico** [comment]
React is not very easy. I've changed the project because I don't succeed with that. I've put the data call in my main app. I think making components speaking is really awful. Now I'm trying to catch some gps value from a component child to compare with my data values with a componentDidUpdate but make an error Maximum call stack size exceeded....

**Comment by DaleWebb** [comment]
You might be invoking an infinite loop with your solution there. What are you trying to do?

**Comment by pinaudnico** [comment]
class GPS extends React.Component {

  componentDidUpdate () {
    this.props.fonctiongps(this.props.coords.latitude);
  }

  state = {
    latitudegps: 0,
    longituedgps: 0
  } 

	render() {
    
  return !this.props.isGeolocationAvailable
      ? <div>Your browser does not support Geolocation</div>
      : !this.props.isGeolocationEnabled
        ? <div>Geolocation is not enabled</div>
        : this.props.coords
          ? <div>
                  <div>Module GPS</div>
                  <div>latitude : {this.props.coords.latitude}
                  <br />
                  longitude : {this.props.coords.longitude}
                  </div>
                  <div><br /></div>
            </div>
          : <div>Getting the location data&hellip; </div>;
          
	}
}


export default geolocated({
  positionOptions: {
    enableHighAccuracy: true,
  },
  userDecisionTimeout: 5000,
})(GPS);

**Comment by pinaudnico** [comment]
it works but done an error

**Comment by pinaudnico** [comment]
oh my divs disappears

**Comment by pinaudnico** [comment]
`import React from 'react';

import {geolocated} from 'react-geolocated';

 

class GPS extends React.Component {

  componentDidUpdate () {
    this.props.fonctiongps(this.props.coords.latitude);
  }

  state = {
    latitudegps: 0,
    longituedgps: 0
  } 

	render() {
    
  return !this.props.isGeolocationAvailable
      ? <div>Your browser does not support Geolocation</div>
      : !this.props.isGeolocationEnabled
        ? <div>Geolocation is not enabled</div>
        : this.props.coords
          ? <div>
                  <div>Module GPS</div>
                  <div>latitude : {this.props.coords.latitude}
                  <br />
                  longitude : {this.props.coords.longitude}
                  </div>
                  <div><br /></div>
            </div>
          : <div>Getting the location data&hellip; </div>;
          
	}
}


export default geolocated({
  positionOptions: {
    enableHighAccuracy: true,
  },
  userDecisionTimeout: 5000,
})(GPS);`

**Comment by pinaudnico** [comment]
> import React from 'react';

import {geolocated} from 'react-geolocated';

 

class GPS extends React.Component {

  componentDidUpdate () {
    this.props.fonctiongps(this.props.coords.latitude);
  }

  state = {
    latitudegps: 0,
    longituedgps: 0
  } 

	render() {
    
  return !this.props.isGeolocationAvailable
      ? <div>Your browser does not support Geolocation</div>
      : !this.props.isGeolocationEnabled
        ? <div>Geolocation is not enabled</div>
        : this.props.coords
          ? <div>
                  <div>Module GPS</div>
                  <div>latitude : {this.props.coords.latitude}
                  <br />
                  longitude : {this.props.coords.longitude}
                  </div>
                  <div><br /></div>
            </div>
          : <div>Getting the location data&hellip; </div>;
          
	}
}


export default geolocated({
  positionOptions: {
    enableHighAccuracy: true,
  },
  userDecisionTimeout: 5000,
})(GPS);

**Comment by pinaudnico** [comment]
how to post with divs ?

**Comment by pinaudnico** [comment]
and the index.js 👍 

import React from 'react';
import ReactDom from 'react-dom';

//components
import Navigation from './navigation'
import Fenetre from './fenetre'

// css
import'../style/index.css';

import GPS from '../pages/GPS'
import Donnees from '../pages/donnees'



class App extends React.Component {
          
state = {
	contenuaffichage: "",
	latitude: "latitude"
}

//call back des fonctions d'affichage
fonctiontexte = () => {
    	this.setState({ contenuaffichage: "texte" });
    	};

fonctionphoto = () => {
    	this.setState({ contenuaffichage: "photo" });
		};

fonctionaudio = () => {
    	this.setState({ contenuaffichage: "audio" });
		};

//call back du module donnees
fonctiongps = (latitude) => {
	this.setState({ latitude: latitude });
	
	};	


render() {
	        
		return (
			<div className="zobi">
				
				<div><Fenetre contenu={this.state.contenuaffichage}/></div>
				<div><Navigation fonctiontexte={this.fonctiontexte}
				                 fonctionphoto={this.fonctionphoto}
				                 fonctionaudio={this.fonctionaudio}/></div>	
                <div><GPS fonctiongps={this.fonctiongps}/></div>
				<div>latitude vu de l'enfant : {this.state.latitude}</div>
				<div></div>

				<div><br /></div>
				<div>Coordonnées enregistrées</div>
				<div>Latitude : {this.props.data.allTableaudespointsCsv.edges[0].node.field2}</div>
				<div>Longitude : {this.props.data.allTableaudespointsCsv.edges[0].node.field3}</div>
			</div>

			)
	}
}


export default App;


export const Query = graphql`
query AppQuery {
allTableaudespointsCsv {
  edges {
	node {
	  field1
	  field2
	  field3
	  field4
	  field5
	  field6
	}
  }
}

}
`

**Comment by jquense** [comment]
@pinaudnico please checkout githubs markdown guide for Syntax highlighting and code blocks: https://guides.github.com/features/mastering-markdown/

its very hard to read your posted code and will be alot easier for folks to help you out if its easier to read :)

**Comment by pinaudnico** [comment]
`import React from 'react';
import ReactDom from 'react-dom';

//components
import Navigation from './navigation'
import Fenetre from './fenetre'

// css
import'../style/index.css';

import GPS from '../pages/GPS'
import Donnees from '../pages/donnees'



class App extends React.Component {
          
state = {
	contenuaffichage: "",
	latitude: "latitude"
}

//call back des fonctions d'affichage
fonctiontexte = () => {
    	this.setState({ contenuaffichage: "texte" });
    	};

fonctionphoto = () => {
    	this.setState({ contenuaffichage: "photo" });
		};

fonctionaudio = () => {
    	this.setState({ contenuaffichage: "audio" });
		};

//call back du module donnees
fonctiongps = (latitude) => {
	this.setState({ latitude: latitude });
	
	};	


render() {
	        
		return (
			<div className="zobi">
				
				<div><Fenetre contenu={this.state.contenuaffichage}/></div>
				<div><Navigation fonctiontexte={this.fonctiontexte}
				                 fonctionphoto={this.fonctionphoto}
				                 fonctionaudio={this.fonctionaudio}/></div>	
                <div><GPS fonctiongps={this.fonctiongps}/></div>
				<div>latitude vu de l'enfant : {this.state.latitude}</div>
				<div></div>

				<div><br /></div>
				<div>Coordonnées enregistrées</div>
				<div>Latitude : {this.props.data.allTableaudespointsCsv.edges[0].node.field2}</div>
				<div>Longitude : {this.props.data.allTableaudespointsCsv.edges[0].node.field3}</div>
			</div>

			)
	}
}


export default App;


export const Query = graphql`
query AppQuery {
allTableaudespointsCsv {
  edges {
	node {
	  field1
	  field2
	  field3
	  field4
	  field5
	  field6
	}
  }
}

}
`






`

**Comment by pinaudnico** [comment]
```
import React from 'react';
import ReactDom from 'react-dom';

//components
import Navigation from './navigation'
import Fenetre from './fenetre'

// css
import'../style/index.css';

import GPS from '../pages/GPS'
import Donnees from '../pages/donnees'



class App extends React.Component {
          
state = {
	contenuaffichage: "",
	latitude: "latitude"
}

//call back des fonctions d'affichage
fonctiontexte = () => {
    	this.setState({ contenuaffichage: "texte" });
    	};

fonctionphoto = () => {
    	this.setState({ contenuaffichage: "photo" });
		};

fonctionaudio = () => {
    	this.setState({ contenuaffichage: "audio" });
		};

//call back du module donnees
fonctiongps = (latitude) => {
	this.setState({ latitude: latitude });
	
	};	


render() {
	        
		return (
			<div className="zobi">
				
				<div><Fenetre contenu={this.state.contenuaffichage}/></div>
				<div><Navigation fonctiontexte={this.fonctiontexte}
				                 fonctionphoto={this.fonctionphoto}
				                 fonctionaudio={this.fonctionaudio}/></div>	
                <div><GPS fonctiongps={this.fonctiongps}/></div>
				<div>latitude vu de l'enfant : {this.state.latitude}</div>
				<div></div>

				<div><br /></div>
				<div>Coordonnées enregistrées</div>
				<div>Latitude : {this.props.data.allTableaudespointsCsv.edges[0].node.field2}</div>
				<div>Longitude : {this.props.data.allTableaudespointsCsv.edges[0].node.field3}</div>
			</div>

			)
	}
}


export default App;


export const Query = graphql`
query AppQuery {
allTableaudespointsCsv {
  edges {
	node {
	  field1
	  field2
	  field3
	  field4
	  field5
	  field6
	}
  }
}

}
`







```

**Comment by pinaudnico** [comment]
```
import React from 'react';

import {geolocated} from 'react-geolocated';

 

class GPS extends React.Component {

  componentDidUpdate () {
    this.props.fonctiongps(this.props.coords.latitude);
  }

  state = {
    latitudegps: 0,
    longituedgps: 0
  } 

	render() {
    
  return !this.props.isGeolocationAvailable
      ? <div>Your browser does not support Geolocation</div>
      : !this.props.isGeolocationEnabled
        ? <div>Geolocation is not enabled</div>
        : this.props.coords
          ? <div>
                  <div>Module GPS</div>
                  <div>latitude : {this.props.coords.latitude}
                  <br />
                  longitude : {this.props.coords.longitude}
                  </div>
                  <div><br /></div>
            </div>
          : <div>Getting the location data&hellip; </div>;
          
	}
}


export default geolocated({
  positionOptions: {
    enableHighAccuracy: true,
  },
  userDecisionTimeout: 5000,
})(GPS);



```

**Comment by pinaudnico** [comment]
ok you see in the last file, componentdidupdate make an error
In the first one, index.js u see as I solve my first problem. Not so beautiful but works.

**Comment by fabien0102** [comment]
@pinaudnico You can add `javascript` after your first \`\`\` to add color in your snippet ;)

Example:
\`\`\`javascript
my code
\`\`\`

**Comment by pinaudnico** [comment]
good thanks

**Comment by pinaudnico** [comment]
```javascript

import React from 'react';
import ReactDom from 'react-dom';

//components
import Navigation from './navigation'
import Fenetre from './fenetre'

// css
import'../style/index.css';

import GPS from '../pages/GPS'
import Donnees from '../pages/donnees'



class App extends React.Component {
          
state = {
	contenuaffichage: "",
	latitude: "latitude"
}

//call back des fonctions d'affichage
fonctiontexte = () => {
    	this.setState({ contenuaffichage: "texte" });
    	};

fonctionphoto = () => {
    	this.setState({ contenuaffichage: "photo" });
		};

fonctionaudio = () => {
    	this.setState({ contenuaffichage: "audio" });
		};

//call back du module donnees
fonctiongps = (latitude) => {
	this.setState({ latitude: latitude });
	
	};	


render() {
	        
		return (
			<div className="zobi">
				
				<div><Fenetre contenu={this.state.contenuaffichage}/></div>
				<div><Navigation fonctiontexte={this.fonctiontexte}
				                 fonctionphoto={this.fonctionphoto}
				                 fonctionaudio={this.fonctionaudio}/></div>	
                <div><GPS fonctiongps={this.fonctiongps}/></div>
				<div>latitude vu de l'enfant : {this.state.latitude}</div>
				<div></div>

				<div><br /></div>
				<div>Coordonnées enregistrées</div>
				<div>Latitude : {this.props.data.allTableaudespointsCsv.edges[0].node.field2}</div>
				<div>Longitude : {this.props.data.allTableaudespointsCsv.edges[0].node.field3}</div>
			</div>

			)
	}
}


export default App;


export const Query = graphql`
query AppQuery {
allTableaudespointsCsv {
  edges {
	node {
	  field1
	  field2
	  field3
	  field4
	  field5
	  field6
	}
  }
}

}
`







```

**Comment by pinaudnico** [comment]
```javascript

import React from 'react';

import {geolocated} from 'react-geolocated';

 

class GPS extends React.Component {

  componentDidUpdate () {
    this.props.fonctiongps(this.props.coords.latitude);
  }

  state = {
    latitudegps: 0,
    longituedgps: 0
  } 

	render() {
    
  return !this.props.isGeolocationAvailable
      ? <div>Your browser does not support Geolocation</div>
      : !this.props.isGeolocationEnabled
        ? <div>Geolocation is not enabled</div>
        : this.props.coords
          ? <div>
                  <div>Module GPS</div>
                  <div>latitude : {this.props.coords.latitude}
                  <br />
                  longitude : {this.props.coords.longitude}
                  </div>
                  <div><br /></div>
            </div>
          : <div>Getting the location data&hellip; </div>;
          
	}
}


export default geolocated({
  positionOptions: {
    enableHighAccuracy: true,
  },
  userDecisionTimeout: 5000,
})(GPS);





```

**Comment by fabien0102** [comment]
And you can edit your post to avoid spamming ;) (top right corner)

**Comment by fabien0102** [comment]
BTW, I don't see any `fonctionGps` in your `props`, you can add `console.log(this)` to understand a little bit more what you have in your react component ;)

**Comment by DaleWebb** [comment]
@pinaudnico I think you should open a new issue for this.

Potentially could be an SSR issue with Gatsby or an error in your application layer logic.

**Comment by pinaudnico** [comment]
ok thanks for all to everyone

**Comment by fabien0102** [comment]
@pinaudnico Or go to reactiflux #need-help :wink: (a discord channel for this kind of case)

https://www.reactiflux.com/

**Comment by pinaudnico** [comment]
yes there is fonctiongps in my index.js 

//call back du module donnees
fonctiongps = (latitude) => {
	this.setState({ latitude: latitude });
	
	};

**Comment by KyleAMathews** [comment]
Seems this conversation has resolved itself — open a new issue please if anyone still needs help!

**Comment by samgermain** [comment]
[Static Queries](https://www.gatsbyjs.org/docs/static-query/) are the way to do this now
